[Go系列教程](https://golangbot.com/learn-golang-series/) [Part30 Error Handling](https://golangbot.com/error-handling/)

# 什么是 errors?

Errors 表明程序中存在异常情况。假设我们试图打开一个文件，而文件系统中不存在这个文件。这是一种异常情况，它以错误的形式表现。

Go 中的错误是个普通的旧值。使用内置类型 `error` 表示错误。

就如同其他类型的内置 [类型type](https://golangbot.com/types/)，比如 int, float64，错误值可以存储在变量中，从函数中返回等等。

现在我们以打开一个不存在的文件为例的程序开始

```go
package main

import (  
    "fmt"
    "os"
)

func main() {  
    f, err := os.Open("/test.txt")
    if err != nil {
        fmt.Println(err)
        return
    }
    fmt.Println(f.Name(), "opened successfully")
}
```

[Run in playground](https://play.golang.org/p/yOhAviFM05)

在上面程序的第9行，我们尝试在路径 `/test.txt` 打开文件。软件包 `os` 的函数 [Open](https://golang.org/pkg/os/#Open) 的签名如下：

` func Open(name string) (file *File, err error)`

*如果文件打开成功，函数 Open 将返回文件句柄，同时错误为 nil，当打开文件出错时，将返回一个非 nil 的错误*。

如果 [函数](https://golangbot.com/functions/) 或 [方法](https://golangbot.com/methods/) 返回错误，依照惯例，它是函数的最后一个返回值。因此，函数 `Open` 最后一个值返回 `err`。

**在Go中处理错误的惯例是将返回值和 `nil` 比较。空值表明没有错误发生，而非空值表示当前有错误。** 在我们的例子中，我们在第10行检查错误是否非空，如果非空，打印错误从主函数返回。

运行该程序将打印：

```shell
open /test.txt: No such file or directory
```

完美，我们在打开一个不存在的文件时获得错误状态。

# 错误类型表示

我们深入一点看看内置类型 `error` 是如何定义的。*error* 是一个有下面定义的 [接口](https://golangbot.com/interfaces-part-1/)类型

```go
type error interface {
  Error() string
}
```

它包含一个签名为 `Error() string` 的方法。任何实现这个接口的类型都可以被用作 error。这个方法提供了错误的描述。

当打印错误时，`fmt.Println` 函数在内部调用 `Error() string` 方法以获取错误的描述。这就是上面程序中的第11行错误描述是如何被打印的。

# 从错误中提供更多信息的不同方式

现在我们知道 `error` 是个接口类型，我们看下如何提取更多关于 error 的信息。

在上面的例子中，我们仅打印 error 的描述。如果我们想要打印导致错误的文件的实际路径该怎么办。一个可能的方法是解析 error 字符串，这是程序的输出

```shell
open /test.txt: No such file or directory
```

**我们可以解析这个错误消息并获得导致文件错误的文件路径 "/text.txt"，但是这是最烂的方式。错误描述可以随时在该语言的较新版本中更改，并且我们的代码将中断。**

是否有一种获取文件名的可靠的方式？当然，标准Go语言库使用不同的方式提供更多关于 errors 的信息。我们来一个一个地看一下。

## 1. 声明基础结构类型并从结构字段中获取更多信息

如果你仔细阅读 [Open](https://golang.org/pkg/os/#OpenFile) 函数的文档，就可以看到它返回一个错误类型 `*PathError`，[PathError](https://golang.org/pkg/os/#PathError)是一个 [struct](https://golangbot.com/structs/)它在标准库中实现如下：

```go
type PathError struct{
  Op   string
  Path string
  Err  error
}

func (e *PathError) Error() string{return e.Op+" "+e.Path+": "+e.Err.Error()}
```

如果你对上面源代码感兴趣，可以点击 https://golang.org/src/os/error.go?s=653:716#L11

从上面的代码，可以理解为 `*PathError` 通过声明 `Error() string` 方法实现了 `error interface`。该方法级联了操作（operation），路径（path）和实际的错误并返回它。这样我们获得错误：

```shell
open /test.txt: No such file or directory
```

结构体 `PathError` 的 `Path` 字段包含了导致错误的文件路径。我们来对上面所写的打印路径做个修改

```go
package main

import (  
    "fmt"
    "os"
)

func main() {  
    f, err := os.Open("/test.txt")
    if err, ok := err.(*os.PathError); ok {
        fmt.Println("File at path", err.Path, "failed to open")
        return
    }
    fmt.Println(f.Name(), "opened successfully")
}
```

在上面程序中，我们使用 [类型断言](https://golangbot.com/interfaces-part-1/#typeassertion)来获得错误接口底层的值。然后使用 `err.Path` 打印路径，这样程序输出

```shell
File at path /test.txt failed to open
```

## 2. 声明基础结构体类型并使用方法获取更多信息

获取更多信息的第2个方式是断言底层的类型，通过调用结构体类型的方法来获取更多信息

通过例子可以更好地理解。

标准库中的 [DNSError](https://golang.org/src/net/net.go?s=17133:17424#L547) 结构体类型如下

```go
type DNSError struct {  
    ...
}

func (e *DNSError) Error() string {  
    ...
}
func (e *DNSError) Timeout() bool {  
    ... 
}
func (e *DNSError) Temporary() bool {  
    ... 
}
```

正如从上面代码中看到的，结构体 `DNSError` 有两个方法 `Timeout() bool` 和 `Temporary() bool` ，他们返回布尔值来表明错误是否是因为超时或临时的原因。

我们写一个断言 `*DNSError` 类型的程序，并调用这些方法来决定错误是临时的还是因为超时。

```go
package main

import (  
    "fmt"
    "net"
)

func main() {  
    addr, err := net.LookupHost("golangbot123.com")
    if err, ok := err.(*net.DNSError); ok {
        if err.Timeout() {
            fmt.Println("operation timed out")
        } else if err.Temporary() {
            fmt.Println("temporary error")
        } else {
            fmt.Println("generic error: ", err)
        }
        return
    }
    fmt.Println(addr)
}
```

*注意：DNS查找不能够在 playground 上运行，请在您本机上运行该程序。*

在上面的程序中，我们尝试获取非法域名 `golangbot123.com` 的 ip 地址。在第10行，我们通过断言类型 `*net.DNSError` 来获得底层的错误值。然后分别查检该错误是否是由于超时或临时产生的。

在我们的例子中，错误既不是由于 `temporary` 也不是由于  `timeout` ，因此程序将输出

```shell
generic error: lookup golangbot123.com: no shch  host
```

如果错误是由于 temporary 或 timeout，相应的 if 语句将会被执行，同时我们可以适当地处理。

## 3. 直接比较

获取错误细节的第三种方式是直接比较 `error` 类型的变量。我们通过例子来理解它。

在 `filepath` 包的 [Glob](https://golang.org/pkg/path/filepath/#Glob) 函数，用于返回所有与模式匹配的文件名字。当模式不正确时，它返回一个 `ErrBadPattern` 的错误。

*ErrBadPattern* 在 `filepath` 包中定义如下：

```go
var ErrBadPattern = errors.New("syntax error in pattern")
```

errors.New() 用于创建一个新的 error。我们将在下一教程详细地讨论它。

当模式是不正确时，Glob 函数返回 `ErrBadPattern`。

我们写个程序检查这个 error

```go
package main

import (  
    "fmt"
    "path/filepath"
)

func main() {  
    files, error := filepath.Glob("[")
    if error != nil && error == filepath.ErrBadPattern {
        fmt.Println(error)
        return
    }
    fmt.Println("matched files", files)
}
```

[Run in playground](https://play.golang.org/p/zbVDDHnMZU)

在上面程序中我们搜索模式为 `[` 的文件，它是不正确的模式。我们检查 error 是否为空。为了获得更多关于 error 的信息，我们直接将它与 `filepath.ErrBadPattern` 比较。如果条件满足，那么是由于错误的格式，打印如下：

```shell
syntax error in pattern
```

标准库使用上述任何一种方法来提供有关错误的更多信息。下一教程我们使用这些方法来创建我们自己的 errors。

# 不要忽略错误

永远不要忽略错误。忽略错误来招来麻烦。下面重写例子，其中列出了所有与模式匹配的文件名称，而忽略了错误处理代码。

```go
package main

import (
	"fmt"
  "path/filepath"
)

func main() {
  files, _ := filepath.Glob("[")
  fmt.Println("matched files", files)
}
```

[Run in playground](https://play.golang.org/p/2k8r_Qg_lc)

我们从上个例子已经知道该模式是无效的。我们使用 `_` 空白标识符忽略 `Glob` 函数返回的错误。我们仅打印匹配的文件，程序输出如下：

```go
matched files []
```

由于忽略了 error，输出看上去没有文件匹配模式，但实际上模式本身是错误的。所以永远不要忽略错误。

在该教程中我们讨论了如何处理发生在程序中的错误，并且如何检查错误来获得更多信息。概括如下：

* 什么是错误？
* 错误的表示
* 从错误中提取更多信息的不种方式
* 不要忽略错误。

**下一教程 - [自定义错误]() **















